### PR TITLE
Refactor observable API implementations

### PR DESCRIPTION
**Pull request by macdrevx**
This is part 2 of 3 toward a fix for #775 (part 1 being #1125).

In this PR, the event listener functionality in MapboxMap and Snapshotter has been extracted into a shared subcomponent named MapboxObservable.

The important aspect here is that it unifies the underlying interaction with MapboxCoreMaps regardless of whether the block-based or Observer-protocol-based API is used to subscribe. Now, in both cases, MapboxObservable manages a wrapper observer that will give us the ability to selectively opt-out of certain notifications in the following PR.

## Pull request checklist:
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Describe the changes in this PR, especially public API changes.
 - [x] Review and agree to the Contributor License Agreement ([CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement)).

### CODE REVIEW

File: Sources/MapboxMaps/Foundation/MapEventsObservable.swift [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+/// :nodoc:
+/// Deprecated. This protocol will be removed from the public API in a future major version.
+public protocol MapEventsObservable: AnyObject {
```
Comment (macdrevx) [inline Sources/MapboxMaps/Foundation/MapEventsObservable.swift:5]: This was just moved to a separate file.

File: Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift [hunk 2]
```diff
@@ -219,4 +229,60 @@ final class MapboxMapTests: XCTestCase {
...
+
+        mapboxMap.onNext(eventType, handler: handlerStub.call(with:))
+
+        XCTAssertEqual(mapboxObservable.onNextStub.invocations.count, 1)
+        XCTAssertEqual(mapboxObservable.onNextStub.invocations.first?.parameters.eventTypes, [eventType])
+        let handler = try XCTUnwrap(mapboxObservable.onNextStub.invocations.first?.parameters.handler)
+        let event = Event(type: "", data: 0)
+        handler(event)
+        XCTAssertEqual(handlerStub.invocations.count, 1)
+        XCTAssertIdentical(handlerStub.invocations.first?.parameters, event)
```
Comment (evil159) [inline Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift:270]: This part looks more like testing out stubbing than testing `MapboxMap`, nuke it?

File: Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift [hunk 3]
```diff
@@ -219,4 +229,60 @@ final class MapboxMapTests: XCTestCase {
...
+
+        mapboxMap.onEvery(eventType, handler: handlerStub.call(with:))
+
+        XCTAssertEqual(mapboxObservable.onEveryStub.invocations.count, 1)
+        XCTAssertEqual(mapboxObservable.onEveryStub.invocations.first?.parameters.eventTypes, [eventType])
+        let handler = try XCTUnwrap(mapboxObservable.onEveryStub.invocations.first?.parameters.handler)
+        let event = Event(type: "", data: 0)
+        handler(event)
+        XCTAssertEqual(handlerStub.invocations.count, 1)
+        XCTAssertIdentical(handlerStub.invocations.first?.parameters, event)
```
Comment (evil159) [inline Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift:286]: Same comment as for the test above ^

File: Tests/MapboxMapsTests/Snapshot/SnapshotterTests.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+        snapshotter.onEvery(eventType, handler: handlerStub.call(with:))
+
+        XCTAssertEqual(mapboxObservable.onEveryStub.invocations.count, 1)
+        XCTAssertEqual(mapboxObservable.onEveryStub.invocations.first?.parameters.eventTypes, [eventType])
+        let handler = try XCTUnwrap(mapboxObservable.onEveryStub.invocations.first?.parameters.handler)
+        let event = Event(type: "", data: 0)
+        handler(event)
+        XCTAssertEqual(handlerStub.invocations.count, 1)
+        XCTAssertIdentical(handlerStub.invocations.first?.parameters, event)
```
Comment (evil159) [inline Tests/MapboxMapsTests/Snapshot/SnapshotterTests.swift:84]: Same comment here and in the test above as in `MapboxMap`.

**Comment by evil159** [comment]
Impressive work! Pretty much nothing to complain about, except maybe naming changes could be considered.

File: Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift (hunk 2 above)
Comment (macdrevx) [reply Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift:270]: The goal of this tests and the others like it is to verify that the handler that's passed to `MapboxMap.onNext` is effectively the same as the one passed to `MapboxObservable.onNext`. If Swift closures were identifiable, we could just compare the one received by MapboxObservable to the one passed to MapboxMap using `===`. Swift doesn't support that though, so instead, we verify that invoking the handler received by MapboxObservable causes the handler passed to MapboxMap to be invoked. This design also allows for the possibility that the handler passed to MapboxObservable is some wrapper around the one passed to MapboxMap. We don't take advantage of that possibility now, but if we did in the future, this test would continue to pass.

Comment (macdrevx) [reply Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift:270]: I'm adding comments to explain this better

File: Sources/MapboxMaps/Foundation/MapboxObservable.swift [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+internal protocol ObservableProtocol: AnyObject {
+    func subscribe(for observer: Observer, events: [String])
+    func unsubscribe(for observer: Observer)
+}
+
+extension Map: ObservableProtocol {}
+
+extension MapSnapshotter: ObservableProtocol {}
+
+internal final class MapboxObservable: MapboxObservableProtocol {
```
[the remark this answers is not in the document]
Comment (macdrevx) [reply Sources/MapboxMaps/Foundation/MapboxObservable.swift:19]: The naming was meant to follow the pattern introduced between MapboxMap and MBMMap. I'll move the `ObservableProtocol` and corresponding extension to a separate file, maybe that will make it more clear.

File: Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift (hunk 2 above)
Comment (evil159) [reply Tests/MapboxMapsTests/Foundation/MapboxMapTests.swift:273]: It is clear now, thanks!

**Review by evil159 — APPROVED** [review]

**Pull request merged**